xla/service/gpu/runtime/nvshmem_api.h

```diff
@@ -0,0 +1,69 @@
+/* Copyright 2024 The OpenXLA Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef XLA_SERVICE_GPU_RUNTIME_NVSHMEM_API_H_
+#define XLA_SERVICE_GPU_RUNTIME_NVSHMEM_API_H_
+
+#include <functional>
+#include <string_view>
+
+#include <cuda.h>
+
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+
+namespace xla::gpu {
+
+//===----------------------------------------------------------------------===//
+// NvshmemApi
+//===----------------------------------------------------------------------===//
+
+class NvshmemApi {
+ public:
+  // Returns a default NvshmemApi for a current process.
+  // NvshmemApi follows the Singleton design pattern
+  static NvshmemApi& Default();
+
+  static void SetEnvInfo(
```

Comment:
`static std::unique_ptr<Collectives> NvshmemApi::Create();` ?

Reply:
Thanks @ezhulenev for reviewing! Are you suggesting we shouldn't use the singleton model and instead require explicit initialization with `NvshmemApi::Create(...)`?

In that case, would we initialize nvshmem during the creation of the gpu client and will the client hold the unique_ptr to it? The problem here is that the NvshmemApi needs to be accesible to the allocator for nvshmem_allocate, and we would need to pipe it through to the nvshmem allocator.

Reply:
Actually it should be a singleton registered similar to NCCL: https://github.com/openxla/xla/blob/cb6562b17ee4eed5aaaa7dd59a37e8303a3b85b0/xla/backends/gpu/collectives/nccl_collectives.cc#L227

Reply:
With the collective registry will it be possible to use both NCCL and NVSHMEM collectives at once? How can we access NvshemCollectives from the registry? It looks like it can only access the one which the highest priority?

Reply:
We can extend the API to get collectives by name if needed. Default one can be the one with highest priority.